cmd/cli/config/set.go

```diff
@@ -26,40 +27,58 @@ func newSetCmd() *cobra.Command {
-			_, err := util.SetupRepoConfig(cmd)
+			// Get the value of the --config flag
+			configFlag, err := cmd.Flags().GetString("config")
 			if err != nil {
-				return fmt.Errorf("failed to setup repo: %w", err)
+				return fmt.Errorf("failed to get config flag: %w", err)
 			}
-			return setConfig(cmd.PersistentFlags().Lookup("config").Value.String(), args[0], args[1:]...)
-		},
-		// provide auto completion for arguments to the `set` command
-		ValidArgsFunction: setAutoComplete,
-	}
 
-	bacalhauCfgDir := "bacalhau"
-	bacalhauCfgFile := config.DefaultFileName
+			// If --config flag is set, use it to set the Viper key
+			if configFlag != "" {
+				viper.Set(cliflags.RootCommandConfigFiles, []string{configFlag})
+			}
 
-	usrCfgDir, err := os.UserConfigDir()
-	if err != nil {
-		log.Warn().Err(err).Msg("Failed to find user-specific configuration directory. Using current directory to write config.")
-	} else {
-		bacalhauCfgDir = filepath.Join(usrCfgDir, bacalhauCfgDir)
-		if err := os.MkdirAll(bacalhauCfgDir, util2.OS_USER_RWX); err != nil {
-			// This means we failed to create a directory either in the current directory, or the user config dir
-			// indicating a some-what serious misconfiguration of the system. We panic here to provide as much
-			// detail as possible.
-			log.Panic().Err(err).Msgf("Failed to create bacalhau configuration directory: %s", bacalhauCfgDir)
-		}
-		bacalhauCfgFile = filepath.Join(bacalhauCfgDir, bacalhauCfgFile)
+			var configPath string
+			// load configs to get the config file path
+			rawConfig, err := util.SetupConfigType(cmd)
+			if err != nil {
+				return fmt.Errorf("failed to setup config: %w", err)
+			}
+
+			configPath = rawConfig.ConfigFileUsed()
+			if configPath == "" {
+				// we fall back to the default config file path $BACALHAU_DIR/config.yaml
+				// this requires initializing a new or opening an existing data-dir
+				bacalhauConfig, err := util.DecodeBacalhauConfig(rawConfig)
+				if err != nil {
+					return fmt.Errorf("failed to decode bacalhau config: %w", err)
+				}
+				_, err = util.SetupRepo(bacalhauConfig)
+				if err != nil {
+					return fmt.Errorf("failed to setup repo: %w", err)
+				}
+				configPath = filepath.Join(bacalhauConfig.DataDir, config.DefaultFileName)
+
+				// create the config file if it doesn't exist
+				if _, err := os.Stat(configPath); os.IsNotExist(err) {
+					if err := os.WriteFile(configPath, []byte{}, util2.OS_USER_RWX); err != nil {
+						return fmt.Errorf("failed to create default config file %s: %w", configPath, err)
+					}
+				}
+			}
+
+			return setConfig(configPath, args[0], args[1:]...)
+		},
+		// Provide auto completion for arguments to the `set` command
+		ValidArgsFunction: setAutoComplete,
 	}
 
-	setCmd.PersistentFlags().String("config", bacalhauCfgFile, "Path to the config file")
+	setCmd.PersistentFlags().String("config", "", "Path to the config file (default is $BACALHAU_DIR/config.yaml)")
 	return setCmd
```

Comment:
This can all be simplified a fair bit. `setConfig` will create the file if it doesn't already exist so no need to create one here.
```suggestion
			// initialize a new or open an existing repo. We need to ensure a repo
			// exists before we can create or modify a config file in it.
			cfg, err := util.SetupRepoConfig(cmd)
			if err != nil {
				return fmt.Errorf("failed to setup repo: %w", err)
			}
			cfgFile := cmd.PersistentFlags().Lookup("config").Value.String()
			if cfgFile == "" {
				cfgFile = filepath.Join(cfg.DataDir, config.DefaultFileName)
			}
			return setConfig(cfgFile, args[0], args[1:]...)
		},
		// provide auto completion for arguments to the `set` command
		ValidArgsFunction: setAutoComplete,
	}

	setCmd.PersistentFlags().String("config", "", "Path to the config file")
	return setCmd
```

Reply:
few issues with this approach:
- a repo is going to be initialized and created unnecessarily even if the user define a config path. The repo can be well different than the one in user's config file as this `--config` flag is different than the one `SetupRepoConfig()` is basing on
- we will create a config instance and log that a config was initialize using paths that can be different from one provided by the user as we are calling `SetupRepoConfig()` before parsing `--config`
- `setConfig` is not creating the config file for us, and we don't want it to. We are only creating a config file if we are using the default path where we are creating config.yaml under the repo. Otherwise if the user provided a config file that doesn't exist, then we better fail instead of creating it to avoid undesired behavour with typos

Here are examples of this behaviour:
```
# config file with datadir setup
→ cat bacalhau_config.yaml
datadir: /Users/walid/test/config

# set a new entry in that config, you will see that repo is getting initialized
# at the default location instead of the one in the config file
→ bacalhau config set api.host 2.2.2.2 --config bacalhau_config.yaml
21:49:11.858 | INF ../../workspace/bacalhau/pkg/config/config.go:211 > Config loaded from: [], and with data-dir /Users/walid/.bacalhau
21:49:11.859 | INF ../../workspace/bacalhau/pkg/repo/fs.go:91 > Initializing repo at /Users/walid/.bacalhau
21:49:12.018 | INF ../../workspace/bacalhau/cmd/cli/config/set.go:49 > Writing config to bacalhau_config.yaml
```

Reply:
Sorry, you're right, the creation of the repo should happen on if a flag isn't provided:
```go
func newSetCmd() *cobra.Command {
	setCmd := &cobra.Command{
		Use:          "set",
		Args:         cobra.MinimumNArgs(2),
		Short:        "Set a value in the config.",
		PreRunE:      hook.ClientPreRunHooks,
		PostRunE:     hook.ClientPostRunHooks,
		SilenceUsage: true,
		RunE: func(cmd *cobra.Command, args []string) error {
			cfgFile := cmd.PersistentFlags().Lookup("config").Value.String()
			if cfgFile == "" {
				// if the user didn't provide a flag, then ensure a repo is initialize and write/create file there.
				cfg, err := util.SetupRepoConfig(cmd)
				if err != nil {
					return err
				}
				cfgFile = filepath.Join(cfg.DataDir, config.DefaultFileName)
			}
			return setConfig(cfgFile, args[0], args[1:]...)
		},
		// provide auto completion for arguments to the `set` command
		ValidArgsFunction: setAutoComplete,
	}

	setCmd.PersistentFlags().String("config", "", "Path to the config file")
	return setCmd
}

func setConfig(cfgFilePath, key string, value ...string) error {
	v := viper.New()
	v.SetConfigFile(cfgFilePath)
	if err := v.ReadInConfig(); err != nil {
		if !os.IsNotExist(err) {
			return err
		}
	}
	parsed, err := types.CastConfigValueForKey(key, value)
	if err != nil {
		return err
	}
	v.Set(key, parsed)
	if err := v.WriteConfigAs(cfgFilePath); err != nil {
		return err
	}

	return nil
}
```

Reply:
As far as `21:49:11.858 | INF ../../workspace/bacalhau/pkg/config/config.go:211 > Config loaded from: [], and with data-dir /Users/walid/.bacalhau`

I think we should only produce this log line when we have loaded a config from a file

Reply:
First, I don't like that we are forking how config path is resolved in `set` command compared to others. I tried to minimize the fork by still calling `SetupConfig` the same as how all other methods do. If in the future we introduce a new logic on how config path is resolved, such as introducing an env variable, we wont fall in bugs where we miss to update the code here as well. It is getting late here. I have an idea to reuse the same config flags that I'll push tomorrow.

Second, we still have to create the config file if it is coming from the repo

Last, I have to disagree about the logging. Even if configs are not loaded from a file because we couldn't fine one or the user didn't provide one, we should still log that to help them understand what is going on

Reply:
Okay, my goal with these comments is to get use to simplify things, my approval is sticky, so merge when you are satisfied.

Reply:
I've made changes here to make sure we are using the same `--config` flag for all commands, both for reading and writing it. Config initialization is now as unified as it can get using `util.SetupConfig()` that relies on that flag

https://github.com/bacalhau-project/bacalhau/pull/4466/commits/b985723b8902376b04dbf08e915ecb397870ef53